patches/nginx-1.11.2-privilege_process.patch

```diff
@@ -0,0 +1,232 @@
+ };
+ 
++static ngx_privilege_worker_ctx_t  ngx_privilege_worker_ctx = {
++    "privilege process"
++};
++
+ 
+ static ngx_cycle_t      ngx_exit_cycle;
+ static ngx_log_t        ngx_exit_log;
+@@ -130,6 +137,7 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
+     ngx_start_worker_processes(cycle, ccf->worker_processes,
+                                NGX_PROCESS_RESPAWN);
+     ngx_start_cache_manager_processes(cycle, 0);
++    ngx_start_privilege_worker_processes(cycle, 0);
+ 
+     ngx_new_binary = 0;
+     delay = 0;
+@@ -224,6 +232,7 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
+                 ngx_start_worker_processes(cycle, ccf->worker_processes,
+                                            NGX_PROCESS_RESPAWN);
+                 ngx_start_cache_manager_processes(cycle, 0);
++                ngx_start_privilege_worker_processes(cycle, 0);
+                 ngx_noaccepting = 0;
+ 
+                 continue;
+@@ -243,6 +252,7 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
+             ngx_start_worker_processes(cycle, ccf->worker_processes,
+                                        NGX_PROCESS_JUST_RESPAWN);
+             ngx_start_cache_manager_processes(cycle, 1);
++            ngx_start_privilege_worker_processes(cycle, 1);
+ 
+             /* allow new processes to start */
+             ngx_msleep(100);
+@@ -257,6 +267,7 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
+             ngx_start_worker_processes(cycle, ccf->worker_processes,
+                                        NGX_PROCESS_RESPAWN);
+             ngx_start_cache_manager_processes(cycle, 0);
++            ngx_start_privilege_worker_processes(cycle, 0);
+             live = 1;
+         }
+ 
+@@ -424,6 +435,34 @@ ngx_start_cache_manager_processes(ngx_cycle_t *cycle, ngx_uint_t respawn)
+ 
+ 
+ static void
++ngx_start_privilege_worker_processes(ngx_cycle_t *cycle, ngx_uint_t respawn)
++{
++    ngx_channel_t    ch;
++    ngx_core_conf_t *ccf;
++
++    ccf = (ngx_core_conf_t *) ngx_get_conf(cycle->conf_ctx,
++                                           ngx_core_module);
++
++    if (!ccf->privilege) {
++        return;
++    }
++
++    ngx_spawn_process(cycle, ngx_privilege_worker_process_cycle,
++                      &ngx_privilege_worker_ctx, ngx_privilege_worker_ctx.name,
++                      respawn ? NGX_PROCESS_JUST_RESPAWN : NGX_PROCESS_RESPAWN);
++
++    ngx_memzero(&ch, sizeof(ngx_channel_t));
++
++    ch.command = NGX_CMD_OPEN_CHANNEL;
++    ch.pid = ngx_processes[ngx_process_slot].pid;
++    ch.slot = ngx_process_slot;
++    ch.fd = ngx_processes[ngx_process_slot].channel[0];
++
++    ngx_pass_open_channel(cycle, &ch);
++}
++
++
++static void
+ ngx_pass_open_channel(ngx_cycle_t *cycle, ngx_channel_t *ch)
+ {
+     ngx_int_t  i;
+@@ -827,7 +866,7 @@ ngx_worker_process_init(ngx_cycle_t *cycle, ngx_int_t worker)
+         }
+     }
+ 
+-    if (geteuid() == 0) {
++    if (ngx_process != NGX_PROCESS_PRIVILEGE && geteuid() == 0) {
```

Comment:
The logic inside is used to modify the process's privilege, the privileged process should have same privilege as master process, it don't need to do this.